File: project_sms_web/js/Student.js

class Student {

  // 생성자(객체)
  constructor(ssn, name, korean, english, math) {

    // this = {};
  
    this.ssn = ssn;
    this.name = name;
    this.korean = korean;
    this.english = english;
    this.math = math;
    // return this;
  
  }


  // 정적(static) 속성 및 메소드 추가
static schoolName = "Ezen 초등학교";

getName(){
  return this.name;
}

// 객체의 생성자 함수(객체)의 메소드를 Student 의 프로토타입 객체에 할당
getSum(){
  return this.korean + this.english + this.math;
}

getEnglish() {
  return this.english;
}

getAverage() {
  return this.getSum() / 3;
}

// Object 프로토타입 객체의 toString() 오버라이딩

toString() {
  return `${this.ssn} \t ${this.name} \t ${this.korean} \t ${this.english} \t ${this.math} \t ${this.getSum()} \t ${this.getAverage()}`
}



}

// console.log(Student);


export {Student};

File: Java_oopV2/src/ezen/oop/Student.java
package ezen.oop;

public class Student<T> /*implements Comparable<Student<Integer>> */{
	private String name;
	private T ssn;

	public Student() {
	}

	public Student(String name, T ssn) {
		super();
		this.name = name;
		this.ssn = ssn;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public T getSsn() {
		return ssn;
	}

	public void setSsn(T ssn) {
		this.ssn = ssn;
	}

	@Override
	public String toString() {
		return "Student [name=" + name + ", ssn=" + ssn + "]";
	}

	@Override
	public boolean equals(Object obj) {
//		System.out.println("hashSet이 equals() 호출함.");
		if(obj instanceof Student) {
			return toString().equals(obj.toString());
		}
		return false;
	}

	@Override
	public int hashCode() {
//		System.out.println("hashSet이 hashCode() 호출함.");
		return toString().hashCode();
	}
	
	
	
	/*
	@Override
	public int compareTo(Student<Integer> student) {
		System.out.println("compareTo 호출됨");
//		if(ssn == student.getSsn()) {
//			return 0;
//			}else if(((Integer)ssn).intValue() < student.getSsn().intValue()) {
//				return -1;
//			}else {
//				return 1;
//			}
			return ((Integer)ssn).intValue() - student.getSsn().intValue();
		}
	*/
	public static void main(String[] args) {
		
		Student<String> student = new Student<>();
		student.setSsn("1234");
		student.getSsn();
		
		Student<Integer> student2 = new Student<>();
		student2.setSsn(1234); // Auto Boxing
		int s = student2.getSsn(); // unboxing
		System.out.println(s);
		
	}




}
